### DISCUSSION TITLE
Clarification on Memory Allocation for Libsodium in C#

### DISCUSSION THREAD

**Discussion by Wojtekxor2** [jedisct1/libsodium#1406] [Q&A]
Hello,

I’m using C# with Libsodium, and I have a couple of questions regarding memory allocation for structs. Currently, I calculate SHA-256 hashes for binary files and also perform encryption using XChaCha20-Poly1305.

I’m a bit unclear on whether I should manually allocate memory for the following structs, or if it's handled automatically:

```
public unsafe struct crypto_secretstream_xchacha20poly1305_state
{
  public fixed byte k[32];
  public fixed byte nonce[12];
  public fixed byte _pad[8];
}

public unsafe struct crypto_hash_sha256_state
{
  public fixed uint state[8];
  public ulong count;
  public fixed byte buf[64];
}
```

Initially, I was declaring them like this:

```
crypto_secretstream_xchacha20poly1305_state state;
LibsodiumAPI.crypto_secretstream_xchacha20poly1305_init_push(&state, headerArg, key);
```

or 

```
crypto_hash_sha256_state st;
LibsodiumAPI.crypto_hash_sha256_init(&st);
```

1. Is this approach correct, or should I be manually allocating memory for these structs using Marshal.AllocHGlobal or another method (e.g., for XChaCha20 specifically or for both)?

2. I came across some Delphi code that aligns memory for XChaCha, and I rewrote it in C#. Here's the code I’m using:

```
public static unsafe crypto_secretstream_xchacha20poly1305_state GetAlignedCryptoGenericHashState(out IntPtr blockStart)
{      
  // Allocate a slightly bigger block keeping the whole block a multiple of 16
  int ALIGN_BYTES = 64;
  int size = sizeof(crypto_secretstream_xchacha20poly1305_state) + ALIGN_BYTES;
  blockStart = Marshal.AllocHGlobal(size);

  // Align the pointer to the nearest multiple of 64
  long alignedPointer = ((long)blockStart + (ALIGN_BYTES - 1)) & ~(ALIGN_BYTES - 1);
  return Marshal.PtrToStructure<crypto_secretstream_xchacha20poly1305_state>((IntPtr)alignedPointer);
}
```
Is this approach necessary? Should I always be aligning the memory for these structs, or is this only required in specific cases?

**Comment by samuel-lucas6** [comment]
Why can't you use an [existing .NET binding](https://doc.libsodium.org/bindings_for_other_languages#bindings-programming-languages)? They save you from having to do this yourself and are heavily tested.

Anyway, if you can't/don't want to use one for whatever reason, you can follow what they're doing, like [this](https://github.com/ektrah/nsec/blob/master/src/Interop/Interop.Hash.Sha256.cs#L51).

**Comment by Wojtekxor2** [comment]
Not that .NET version I need it for. I looked at nSec however I didn't find chacha implementation for long data (with update method) ...

**Reply by samuel-lucas6** [reply]
Which version of .NET are you using? [My libsodium binding](https://www.geralt.xyz/authenticated-encryption/stream-and-file-encryption) supports secretstream, but I only support the latest versions of .NET as well.
